### ISSUE TITLE
evaluate result is #VALUE! when the formula range >=256

### ISSUE THREAD

**Issue by alexshen-sx** [bradbase/xlcalculator#14]
Hi bradbase

I'm new to python and just learning how to use xlcalculator to parse formulas from xlsx file. 
I might found a bug that the evaluate result is #VALUE! when the formula range >= 256.
It could be reproduced by:

there're 2 sheets in workbook test.xlsx named sheet1 and sheet2
there're some random numbers in sheet2.A1:AN
the formula of shee1.A1 is =AVERAGE(sheet2!A1:AN)

I'm using below code to get value from xlsx, it works when N<256, but the result would be #VALUE! once N >=256.

filename = r'test.xlsx'
compiler = ModelCompiler()
new_model = compiler.read_and_parse_archive(filename)
evaluator = Evaluator(new_model)
value = evaluator.evaluate('sheet1!A1')
print(value)

Best Regards,
Alex

**Comment by bradbase** [comment]
Hi Alex,

Thanks for having a go at Python with xlcalculator.

And thanks for the bug report. I won't be able to get to this for a couple of days but that sounds like a really bizarre limitation. There is nothing obvious that I can think of that would cause such a thing.

For your information the functions are currently in a separate library called xlfunctions. Sometime ("soon"?) I will roll these functions into xlcalculator. They were in xlcalculator originally but I made an error in judgement and separated them. Average is in statistics.py. You are welcome to go there and see the code that actually performs the averaging. It doesn't have a constraint.

Cheers,

Brad

**Comment by strichter** [comment]
I could not reproduce this error. 

I have used the attached XLSX:
[test.xlsx](https://github.com/bradbase/xlcalculator/files/5011220/test.xlsx)

I used the following script named `test.py`:

```
from xlcalculator import *

filename = 'test.xlsx'
compiler = ModelCompiler()
new_model = compiler.read_and_parse_archive(filename)
evaluator = Evaluator(new_model)
value = evaluator.evaluate('Sheet1!A1')
print(value)
```

I used the following command:
```
xlcalculator$ ve/bin/python test.py
200.5
```

**Comment by alexshen-sx** [comment]
> Hi Alex,
> 
> Thanks for having a go at Python with xlcalculator.
> 
> And thanks for the bug report. I won't be able to get to this for a couple of days but that sounds like a really bizarre limitation. There is nothing obvious that I can think of that would cause such a thing.
> 
> For your information the functions are currently in a separate library called xlfunctions. Sometime ("soon"?) I will roll these functions into xlcalculator. They were in xlcalculator originally but I made an error in judgement and separated them. Average is in statistics.py. You are welcome to go there and see the code that actually performs the averaging. It doesn't have a constraint.
> 
> Cheers,
> 
> Brad

Thanks Brad for quick response and sorry for my incorrect statements. I'm using COUNTA in xlsx instead of AVERAGE, I'll describe more details on how to reproduce this.

**Comment by alexshen-sx** [comment]
> I could not reproduce this error.
> 
> I have used the attached XLSX:
> [test.xlsx](https://github.com/bradbase/xlcalculator/files/5011220/test.xlsx)
> 
> I used the following script named `test.py`:
> 
> ```
> from xlcalculator import *
> 
> filename = 'test.xlsx'
> compiler = ModelCompiler()
> new_model = compiler.read_and_parse_archive(filename)
> evaluator = Evaluator(new_model)
> value = evaluator.evaluate('Sheet1!A1')
> print(value)
> ```
> 
> I used the following command:
> 
> ```
> xlcalculator$ ve/bin/python test.py
> 200.5
> ```

Hi Strichter

Sorry for my mistake about the AVERAGE statements, I'm actually using COUNTA.

Here's the actual xlsx I'm using.
shee1!A1 = AVERAGE(sheet2!.A1:A256)/256
shee1!B1 = COUNTA(sheet2!.A1:A256)/256

The value get back from A1 is correct but shows #VALUE! from B1.
However if change the range to A1:A255, then the result for B1 would be good too.

Best Regards,
Alex

**Comment by bradbase** [comment]
Hi @alexshen-sx,

There could be two bugs in our implementation of COUNTA. Certainly the argument count is in error, but pedantically there may be a secondary bug in the requirement for the function.

The Microsoft documentation specifies the COUNTA function requires 1 argument and has up to 255 optional arguments, totaling 256.

xlfunctions statistics.py:48

Currently COUNTA limits all arguments to a count of 255 and, on my first reading of the code, does not appear to require a first argument.

The Microsoft documentation.

SYNTAX:
COUNTA(value1, [value2], ...)

The COUNTA function syntax has the following arguments:
- value1    Required. The first argument representing the values that you want to count.
- value2, ...    Optional. Additional arguments representing the values that you want to count, up to a maximum of 255 arguments.

https://support.office.com/en-us/article/counta-function-7dc98875-d5c1-46f1-9a82-53f3219e2509

**Comment by bradbase** [comment]
I can't act upon this until Friday. Most likely a fix published Sunday.

**Comment by alexshen-sx** [comment]
Thanks Brad, there might be another issue when involving with AVERAGE.
If the range contains empty cell, Excel works well but xlcalculator won't be able to evaluate the result and the output would be #VALUE!.

Best Regards,
Alex

**Comment by bradbase** [comment]
Hi @alexshen-sx,

I have finally got some time to put into xlcalculator. I've in-boarded xlfunctions, currently in a branch, which will assist everything to do with xlcalculator.

I'll now progress things WRT the COUNTA function.

**Comment by bradbase** [comment]
I've had a look at the COUNTA tests. We already test for missing first argument so that's not an issue.

I have drafted a fix and added three test cases for COUNTA in the tests that compare xlcalculator evaluation against Excel's results;
- range which contains 255 cells (result = 255)
- range which contains 256 cells (result = 256)
- a range which contains 256 cells where one is empty (result = 255)

This has been developed against the in-boarded xlfunctions code so is not yet part of a release.

**Comment by bradbase** [comment]
This fix has been released in 0.2.3
